## Overview

In this PA, you will will be creating and manipulating basic objects to uncover a top secret message.

### Hint 1

As you work through this PA, you will encounter some code that does not work as you want it to.  Don't despair!  Errors (when R is unable to run your code) and bugs (when the code doesn't do what you hoped) are a natural part of coding.  Even the best of the best deal with these problems regularly - learning to track down the issue is a skill that you can learn and practice.

Google is your friend!  If an error is confusing, copy it into Google and see what other people are saying.  If you don't know how to do something, search for it.

### Hint 2

Just because there is no error message doesn't mean everything went smoothly.  You should be regularly printing out your objects, to see they contain the information and format that you expect.

For example, if you have an object named `my_thing`, you might first want to simply look at the object by typing its name.

However, you might also want to make sure it is the right **object type** and/or the right **object structure**.


In [1]:
my_thing = "my_thing"
len(my_thing)

8

## Part One: Setup

Each of the following code chunks will cause an error and/or do the desired task incorrectly.  Find the mistake, and correct it to complete the intended action.




1. Create lists containing the upper case letters, lower case letters, and some punctuation marks.


In [2]:
lower_case = ["a", "b", "c", "d", "e", "f", "g", "h", "i", "j", "k", "l", "m", "n", "o", "p", "q", "r", "s", "t", "u", "v", "w", "x", "y", "z"]

upper_case = ["A", "B", "C", "D", "E", "F", "G", "H" "I", "J", "K", "L", "M", "N", "O", "P", "Q", "R", "S", "T", "U", "V", "W", "X", "Y", "Z"]

punctuation = [".", ",", "!", "?", "'", '"', "(", ")", " ", "-", ";", ":"]
print(lower_case)
print(upper_case)
print(punctuation)

['a', 'b', 'c', 'd', 'e', 'f', 'g', 'h', 'i', 'j', 'k', 'l', 'm', 'n', 'o', 'p', 'q', 'r', 's', 't', 'u', 'v', 'w', 'x', 'y', 'z']
['A', 'B', 'C', 'D', 'E', 'F', 'G', 'HI', 'J', 'K', 'L', 'M', 'N', 'O', 'P', 'Q', 'R', 'S', 'T', 'U', 'V', 'W', 'X', 'Y', 'Z']
['.', ',', '!', '?', "'", '"', '(', ')', ' ', '-', ';', ':']
['a', 'b', 'c', 'd', 'e', 'f', 'g', 'h', 'i', 'j', 'k', 'l', 'm', 'n', 'o', 'p', 'q', 'r', 's', 't', 'u', 'v', 'w', 'x', 'y', 'z']
['A', 'B', 'C', 'D', 'E', 'F', 'G', 'HI', 'J', 'K', 'L', 'M', 'N', 'O', 'P', 'Q', 'R', 'S', 'T', 'U', 'V', 'W', 'X', 'Y', 'Z']
['.', ',', '!', '?', "'", '"', '(', ')', ' ', '-', ';', ':']


2. Make one long list containing all the symbols.  Find the length of this list.


In [3]:
all_symbols = (lower_case + upper_case + punctuation)
len(all_symbols)

63

3. Repeat the previous steps, but instead of using *list* type objects, use **numpy arrays**.  Don't forget to **import** the numpy library!


In [4]:
import numpy as np

lower_case = np.array(["a", "b", "c", "d", "e", "f", "g", "h", "i", "j", "k", "l", "m", "n", "o", "p", "q", "r", "s", "t", "u", "v", "w", "x", "y", "z"])

upper_case = np.array(["A", "B", "C", "D", "E", "F", "G", "H", "I", "J", "K", "L", "M", "N", "O", "P", "Q", "R", "S", "T", "U", "V", "W", "X", "Y", "Z"])

punctuation = np.array([".", ",", "!", "?", "'", "\"", "(", ")", " ", "-", ";", ":"])

## Use np.concatenate to combine into long array
all_symbols = np.concatenate((lower_case, upper_case, punctuation))

len(all_symbols)


64

4. Make a **pandas data frame**, with one column called "Symbol" containing the `all_symbols` array.


In [5]:
import pandas as pd

symbols_df = pd.DataFrame((all_symbols))

symbols_df.rename(columns={0: 'Symbol'}, inplace=True)
print(symbols_df)

   Symbol
0       a
1       b
2       c
3       d
4       e
..    ...
59      )
60       
61      -
62      ;
63      :

[64 rows x 1 columns]
   Symbol
0       a
1       b
2       c
3       d
4       e
..    ...
59      )
60       
61      -
62      ;
63      :

[64 rows x 1 columns]


## Part Two: Decoding the secret message.


In [6]:
top_secret = pd.read_csv("https://www.dropbox.com/s/k72h1zewk4gtqep/PA_Secret_Code?dl=1", header=None)

top_secret1 = np.array(top_secret[0].to_numpy())
for i in range(len(top_secret1)): #i is the index of the list
    top_secret1[i] = (((top_secret1[i] + 14) * 18) - 257)
    top_secret1[i] = (np.exp(top_secret1[i]))
    top_secret1[i] = (top_secret1[i]**2)
    print(top_secret1[i])





59.00000000000038
50.0000000000018
28.999999999999606
8.999999999999957
28.999999999999606
73.00000000000169
16.00000000000012
34.000000000000156
27.00000000000134
31.999999999998476
34.000000000000156
33.000000000000696
33.000000000000696
82.00000000000198
90.99999999999498
33.000000000000696
47.99999999999752
39.999999999999766
98.99999999999622
79.00000000000091
50.0000000000018
48.999999999998096
107.00000000000347
66.9999999999973
62.000000000001165
60.9999999999993
93.00000000000136
64.00000000000071
82.99999999999987
121.00000000000563
100.99999999999564
79.00000000000091
88.00000000000254
73.00000000000169
89.00000000000075
132.9999999999948
134.9999999999975
3797.000000000161
978.000000000004
160.999999999998
98.00000000000198
3804.9999999999104
134.9999999999975
712.9999999999766
413.9999999999783
116.99999999999525
3815.0000000000778
97.00000000000068
293.99999999998846
115.99999999999883
3822.999999999989
153.0000000000084
187.00000000000486
276.9999999999973
113.9999999999

##Checkpoint 1

In [7]:
np.set_printoptions(suppress=True)
#reshape to 5 columns in Fortran order
five_column_topsecret = top_secret1.reshape(-1, 5, order = 'F')
print(five_column_topsecret)

[[  59.   144.   361.     9.    20.5]
 [  50.     1.  1521.     7.5   30.5]
 [  29.   529.   225.    10.5    3. ]
 [   9.   361.   225.     3.5    9. ]
 [  29.  3721.    16.     4.     0.5]
 [  73.  1521.  3721.    30.5    1. ]
 [  16.    64.     1.     0.5    5. ]
 [  34.     1.   529.     7.     7.5]
 [  27.  1521.    64.     2.    10.5]
 [  32.  3721.    81.    30.5    9.5]
 [  34.     9.   144.    19.5   30.5]
 [  33.     1.    25.     4.     2. ]
 [  33.  1521.  3721.     9.     0.5]
 [  82.     9.    81.     7.5   12.5]
 [  91.    64.   196.    10.5   27.5]
 [  33.  3025.  3721.     3.5   30.5]
 [  48.  3721.  1521.     4.    14.5]
 [  40.   784.   164.    30.5    0.5]
 [  99.    25.   325.    30.5    6. ]
 [  79.   529.   541.    30.5    6. ]
 [  50.     1.   149.    23.     7.5]
 [  49.   324.   164.     4.     7.5]
 [ 107.    25.  1621.     2.5    4. ]
 [  67.  3721.  2909.    30.5   27.5]
 [  62.  1521.  3721.    11.    30.5]
 [  61.    64.   729.     7.5   14.5]
 [  93.    2

In [8]:
under_17 = np.array(five_column_topsecret[five_column_topsecret < 17])
below_17_count = np.sum(five_column_topsecret < 17)
print(below_17_count)

352
352


In [9]:
#under_17_5_cols = np.pad(under_17, (0, 5 - len(under_17) % 5), 'constant')
#reshape_under_17_5_cols = under_17_5_cols.reshape(-1, 5, order = 'F')
#print(reshape_under_17_5_cols)
#print(reshape_under_17_5_cols.shape[0])

In [10]:
even = np.arange(0, 383, 2)
evens = even[even != 0]
evens_array = np.array(evens)

print(evens_array)

[  2   4   6   8  10  12  14  16  18  20  22  24  26  28  30  32  34  36  38
  40  42  44  46  48  50  52  54  56  58  60  62  64  66  68  70  72  74  76
  78  80  82  84  86  88  90  92  94  96  98 100 102 104 106 108 110 112 114
 116 118 120 122 124 126 128 130 132 134 136 138 140 142 144 146 148 150 152
 154 156 158 160 162 164 166 168 170 172 174 176 178 180 182 184 186 188 190
 192 194 196 198 200 202 204 206 208 210 212 214 216 218 220 222 224 226 228
 230 232 234 236 238 240 242 244 246 248 250 252 254 256 258 260 262 264 266
 268 270 272 274 276 278 280 282 284 286 288 290 292 294 296 298 300 302 304
 306 308 310 312 314 316 318 320 322 324 326 328 330 332 334 336 338 340 342
 344 346 348 350 352 354 356 358 360 362 364 366 368 370 372 374 376 378 380
 382]
[  2   4   6   8  10  12  14  16  18  20  22  24  26  28  30  32  34  36  38
  40  42  44  46  48  50  52  54  56  58  60  62  64  66  68  70  72  74  76
  78  80  82  84  86  88  90  92  94  96  98 100 102 104 106 108 110 1

In [11]:
evens_vector = evens_array[:].reshape(-1, 1)
print(evens_vector)

[[  2]
 [  4]
 [  6]
 [  8]
 [ 10]
 [ 12]
 [ 14]
 [ 16]
 [ 18]
 [ 20]
 [ 22]
 [ 24]
 [ 26]
 [ 28]
 [ 30]
 [ 32]
 [ 34]
 [ 36]
 [ 38]
 [ 40]
 [ 42]
 [ 44]
 [ 46]
 [ 48]
 [ 50]
 [ 52]
 [ 54]
 [ 56]
 [ 58]
 [ 60]
 [ 62]
 [ 64]
 [ 66]
 [ 68]
 [ 70]
 [ 72]
 [ 74]
 [ 76]
 [ 78]
 [ 80]
 [ 82]
 [ 84]
 [ 86]
 [ 88]
 [ 90]
 [ 92]
 [ 94]
 [ 96]
 [ 98]
 [100]
 [102]
 [104]
 [106]
 [108]
 [110]
 [112]
 [114]
 [116]
 [118]
 [120]
 [122]
 [124]
 [126]
 [128]
 [130]
 [132]
 [134]
 [136]
 [138]
 [140]
 [142]
 [144]
 [146]
 [148]
 [150]
 [152]
 [154]
 [156]
 [158]
 [160]
 [162]
 [164]
 [166]
 [168]
 [170]
 [172]
 [174]
 [176]
 [178]
 [180]
 [182]
 [184]
 [186]
 [188]
 [190]
 [192]
 [194]
 [196]
 [198]
 [200]
 [202]
 [204]
 [206]
 [208]
 [210]
 [212]
 [214]
 [216]
 [218]
 [220]
 [222]
 [224]
 [226]
 [228]
 [230]
 [232]
 [234]
 [236]
 [238]
 [240]
 [242]
 [244]
 [246]
 [248]
 [250]
 [252]
 [254]
 [256]
 [258]
 [260]
 [262]
 [264]
 [266]
 [268]
 [270]
 [272]
 [274]
 [276]
 [278]
 [280]
 [282]
 [284]
 [286]

In [12]:
step_7 = five_column_topsecret[:, 0] - evens_vector
print(step_7)

[[  57.   48.   27. ...  401. 4099.  389.]
 [  55.   46.   25. ...  399. 4097.  387.]
 [  53.   44.   23. ...  397. 4095.  385.]
 ...
 [-319. -328. -349. ...   25. 3723.   13.]
 [-321. -330. -351. ...   23. 3721.   11.]
 [-323. -332. -353. ...   21. 3719.    9.]]
[[  57.   48.   27. ...  401. 4099.  389.]
 [  55.   46.   25. ...  399. 4097.  387.]
 [  53.   44.   23. ...  397. 4095.  385.]
 ...
 [-319. -328. -349. ...   25. 3723.   13.]
 [-321. -330. -351. ...   23. 3721.   11.]
 [-323. -332. -353. ...   21. 3719.    9.]]


In [13]:
step_8 = five_column_topsecret[17:24, 3] - 100
print(step_8)

[-69.5 -69.5 -69.5 -77.  -96.  -97.5 -69.5]
[-69.5 -69.5 -69.5 -77.  -96.  -97.5 -69.5]


In [14]:
step_9 = five_column_topsecret[:, 3:5] * 2
print(step_9)

[[18. 41.]
 [15. 61.]
 [21.  6.]
 [ 7. 18.]
 [ 8.  1.]
 [61.  2.]
 [ 1. 10.]
 [14. 15.]
 [ 4. 21.]
 [61. 19.]
 [39. 61.]
 [ 8.  4.]
 [18.  1.]
 [15. 25.]
 [21. 55.]
 [ 7. 61.]
 [ 8. 29.]
 [61.  1.]
 [61. 12.]
 [61. 12.]
 [46. 15.]
 [ 8. 15.]
 [ 5.  8.]
 [61. 55.]
 [22. 61.]
 [15. 29.]
 [18.  1.]
 [16. 12.]
 [ 1. 12.]
 [12.  1.]
 [61. 25.]
 [ 2. 55.]
 [12. 58.]
 [ 1. 61.]
 [ 4. 61.]
 [ 5. 61.]
 [61. 34.]
 [23.  5.]
 [ 5. 61.]
 [14.  3.]
 [39.  8.]
 [61. 15.]
 [19. 18.]
 [14. 39.]
 [ 9. 12.]
 [ 3.  5.]
 [11.  4.]
 [ 5. 61.]
 [18.  9.]
 [62. 14.]
 [19. 61.]
 [14.  8.]
 [ 1.  9.]
 [ 3. 19.]
 [11. 61.]
 [55. 10.]
 [61. 15.]
 [34. 25.]
 [ 5. 53.]
 [61. 61.]
 [12. 57.]
 [ 5. 46.]
 [ 6. 23.]
 [39.  1.]
 [61. 19.]
 [ 9. 61.]
 [39.  2.]
 [61. 18.]
 [ 4.  9.]
 [ 5. 12.]
 [ 1. 12.]
 [ 4.  9.]
 [54.  7.]
 [61. 54.]
 [ 1. 61.]
 [14.  1.]
 [ 4. 14.]
 [61.  4.]
 [23. 61.]
 [ 9. 39.]
 [39.  8.]
 [ 8.  5.]
 [61. 61.]
 [ 9. 19.]
 [39. 12.]
 [19.  9.]
 [61. 39.]
 [ 8.  8.]
 [ 5. 25.]
 [ 1. 61.]
 [ 4. 39.]

In [15]:
np.set_printoptions(suppress=True)
five_column_topsecret[:, 0] = five_column_topsecret[:, 0] - evens_vector.reshape(-1)
five_column_topsecret[17:24, 3] = five_column_topsecret[17:24, 3] - 100
five_column_topsecret[:, 3:5] = five_column_topsecret[:, 3:5] * 2
print(five_column_topsecret)

[[  57.  144.  361.   18.   41.]
 [  46.    1. 1521.   15.   61.]
 [  23.  529.  225.   21.    6.]
 [   1.  361.  225.    7.   18.]
 [  19. 3721.   16.    8.    1.]
 [  61. 1521. 3721.   61.    2.]
 [   2.   64.    1.    1.   10.]
 [  18.    1.  529.   14.   15.]
 [   9. 1521.   64.    4.   21.]
 [  12. 3721.   81.   61.   19.]
 [  12.    9.  144.   39.   61.]
 [   9.    1.   25.    8.    4.]
 [   7. 1521. 3721.   18.    1.]
 [  54.    9.   81.   15.   25.]
 [  61.   64.  196.   21.   55.]
 [   1. 3025. 3721.    7.   61.]
 [  14. 3721. 1521.    8.   29.]
 [   4.  784.  164. -139.    1.]
 [  61.   25.  325. -139.   12.]
 [  39.  529.  541. -139.   12.]
 [   8.    1.  149. -154.   15.]
 [   5.  324.  164. -192.   15.]
 [  61.   25. 1621. -195.    8.]
 [  19. 3721. 2909. -139.   55.]
 [  12. 1521. 3721.   22.   61.]
 [   9.   64.  729.   15.   29.]
 [  39.   25.  196.   18.    1.]
 [   8. 3721.   16.   16.   12.]
 [  25. 1296. 2916.    1.   12.]
 [  61.  441. 3721.   12.    1.]
 [  39.   

In [16]:
step_10 = np.reshape(five_column_topsecret, (-1,1), order = 'F')
print(step_10)

[[  57.]
 [  46.]
 [  23.]
 [   1.]
 [  19.]
 [  61.]
 [   2.]
 [  18.]
 [   9.]
 [  12.]
 [  12.]
 [   9.]
 [   7.]
 [  54.]
 [  61.]
 [   1.]
 [  14.]
 [   4.]
 [  61.]
 [  39.]
 [   8.]
 [   5.]
 [  61.]
 [  19.]
 [  12.]
 [   9.]
 [  39.]
 [   8.]
 [  25.]
 [  61.]
 [  39.]
 [  15.]
 [  22.]
 [   5.]
 [  19.]
 [  61.]
 [  61.]
 [3721.]
 [ 900.]
 [  81.]
 [  16.]
 [3721.]
 [  49.]
 [ 625.]
 [ 324.]
 [  25.]
 [3721.]
 [   1.]
 [ 196.]
 [  16.]
 [3721.]
 [  49.]
 [  81.]
 [ 169.]
 [   4.]
 [ 144.]
 [  25.]
 [3721.]
 [  81.]
 [ 196.]
 [3721.]
 [1521.]
 [  64.]
 [  25.]
 [3721.]
 [ 529.]
 [   1.]
 [   4.]
 [  25.]
 [4096.]
 [3721.]
 [ 729.]
 [ 144.]
 [ 144.]
 [3721.]
 [ 169.]
 [  81.]
 [ 169.]
 [ 361.]
 [ 625.]
 [3721.]
 [ 529.]
 [  25.]
 [ 324.]
 [  25.]
 [3721.]
 [1521.]
 [  64.]
 [  25.]
 [3721.]
 [   4.]
 [ 225.]
 [ 324.]
 [ 225.]
 [  49.]
 [ 225.]
 [ 484.]
 [  25.]
 [ 361.]
 [2916.]
 [3721.]
 [3721.]
 [3721.]
 [ 729.]
 [ 196.]
 [  16.]
 [3721.]
 [1521.]
 [  64.]
 [  25.]
 [3721.]
 

##Checkpoint 2

In [17]:
print(len(step_10[500:]))
check2 = np.all(step_10[500:] < 100)
print(check2)

455
True
455
True


In [18]:
#step 11
step_10[37:465] = np.sqrt(step_10[37:465])


In [19]:
#step 12
step_10 = step_10.round(decimals=0)
print(step_10)

[[  57.]
 [  46.]
 [  23.]
 [   1.]
 [  19.]
 [  61.]
 [   2.]
 [  18.]
 [   9.]
 [  12.]
 [  12.]
 [   9.]
 [   7.]
 [  54.]
 [  61.]
 [   1.]
 [  14.]
 [   4.]
 [  61.]
 [  39.]
 [   8.]
 [   5.]
 [  61.]
 [  19.]
 [  12.]
 [   9.]
 [  39.]
 [   8.]
 [  25.]
 [  61.]
 [  39.]
 [  15.]
 [  22.]
 [   5.]
 [  19.]
 [  61.]
 [  61.]
 [  61.]
 [  30.]
 [   9.]
 [   4.]
 [  61.]
 [   7.]
 [  25.]
 [  18.]
 [   5.]
 [  61.]
 [   1.]
 [  14.]
 [   4.]
 [  61.]
 [   7.]
 [   9.]
 [  13.]
 [   2.]
 [  12.]
 [   5.]
 [  61.]
 [   9.]
 [  14.]
 [  61.]
 [  39.]
 [   8.]
 [   5.]
 [  61.]
 [  23.]
 [   1.]
 [   2.]
 [   5.]
 [  64.]
 [  61.]
 [  27.]
 [  12.]
 [  12.]
 [  61.]
 [  13.]
 [   9.]
 [  13.]
 [  19.]
 [  25.]
 [  61.]
 [  23.]
 [   5.]
 [  18.]
 [   5.]
 [  61.]
 [  39.]
 [   8.]
 [   5.]
 [  61.]
 [   2.]
 [  15.]
 [  18.]
 [  15.]
 [   7.]
 [  15.]
 [  22.]
 [   5.]
 [  19.]
 [  54.]
 [  61.]
 [  61.]
 [  61.]
 [  27.]
 [  14.]
 [   4.]
 [  61.]
 [  39.]
 [   8.]
 [   5.]
 [  61.]
 

In [20]:
#step 13
step_10[step_10 == 39] = 20
print(step_10)

[[  57.]
 [  46.]
 [  23.]
 [   1.]
 [  19.]
 [  61.]
 [   2.]
 [  18.]
 [   9.]
 [  12.]
 [  12.]
 [   9.]
 [   7.]
 [  54.]
 [  61.]
 [   1.]
 [  14.]
 [   4.]
 [  61.]
 [  20.]
 [   8.]
 [   5.]
 [  61.]
 [  19.]
 [  12.]
 [   9.]
 [  20.]
 [   8.]
 [  25.]
 [  61.]
 [  20.]
 [  15.]
 [  22.]
 [   5.]
 [  19.]
 [  61.]
 [  61.]
 [  61.]
 [  30.]
 [   9.]
 [   4.]
 [  61.]
 [   7.]
 [  25.]
 [  18.]
 [   5.]
 [  61.]
 [   1.]
 [  14.]
 [   4.]
 [  61.]
 [   7.]
 [   9.]
 [  13.]
 [   2.]
 [  12.]
 [   5.]
 [  61.]
 [   9.]
 [  14.]
 [  61.]
 [  20.]
 [   8.]
 [   5.]
 [  61.]
 [  23.]
 [   1.]
 [   2.]
 [   5.]
 [  64.]
 [  61.]
 [  27.]
 [  12.]
 [  12.]
 [  61.]
 [  13.]
 [   9.]
 [  13.]
 [  19.]
 [  25.]
 [  61.]
 [  23.]
 [   5.]
 [  18.]
 [   5.]
 [  61.]
 [  20.]
 [   8.]
 [   5.]
 [  61.]
 [   2.]
 [  15.]
 [  18.]
 [  15.]
 [   7.]
 [  15.]
 [  22.]
 [   5.]
 [  19.]
 [  54.]
 [  61.]
 [  61.]
 [  61.]
 [  27.]
 [  14.]
 [   4.]
 [  61.]
 [  20.]
 [   8.]
 [   5.]
 [  61.]
 

##Checkpoint 3


In [21]:
even_count = np.sum(step_10 % 2 == 0)
print(even_count)


345
345


### Hint

Recall that basic math operations are **vectorized**.  This means you don't have to repeat the process for each item in a list; you can apply the math to the whole thing at once.

### Instructions

By altering this top secret set of numbers, you will be able to create a message.  Write your own code to complete the steps below.

1. Add 14 to every number.
2. Multiply every number by 18, then subtract 257.
3. Exponentiate every number.  (That is, do e^[number].  You may have to Google how to do this!)
4. Square every number.

**Checkpoint:** Headquarters has informed you that at this stage of decoding, there should be 352 numbers in the secret message that are below 17.  Write code to check if this is true for your data.


5. Turn your one-dimensional array of numbers into a two-dimensional array (matrix) with 5 columns. This conversion should be **column-wise**; that is, the first several elements of the array should all end up in the first **column** of your matrix.
6. Separately from your top secret numbers, create an array of all the even numbers between 1 and 382. Name it "evens". That is, "evens" should contain 2, 4, 6, 8 ..., 382.
7. Subtract the "evens" vector from the first column of your secret message matrix.
8. Subtract 100 from all numbers in the 18-24th rows of the 3rd column.
9. Multiply all numbers in the 4th and 5th column by 2.
10. Turn your matrix back into a one dimensional array. Again this should be **column-wise**, so that the first column of the matrix becomes the first elements of the array.


**Checkpoint:** Headquarters has informed you that at this stage of decoding, all numbers in indices 500 and beyond are below 100.  Write code to check if your numbers fulfill this requirement.

11. Take the square root of all numbers in indices 37 to 464.
12. Round all numbers to the nearest whole number.
13. Replace all instances of the number 39 with 20.

**Checkpoint:** Headquarters has informed you that your final message should have 344 even numbers.  Write code to check if your numbers fulfill this requirement.

## Part 3: The secret message!

Run the following to discover the final message!  Google the first line of this message, if you do not recognize it, to see what it is.


In [49]:
step_10 = step_10.flatten()
valid_indices = np.clip(step_10 -1, 0, len(symbols_df) - 1) #extra step solution with assistant from Gemini: ensures all indices are valid due to 5 invalid negative indices
''.join(map(str, symbols_df.loc[valid_indices]['Symbol']))

'\'Twas brillig, and the slithy toves   Did gyre and gimble in the wabe: All mimsy were the borogoves,   And the mome raths outgrabe. "Beware the Jabberwock, my son!   The jaws that bite, the claws that catch! Beware the Jubjub bird, and shun   The frumious Bandersnatch!" He took his vorpal sword in hand;   Long time the manxome foe he sought- So rested he by the Tumtum tree   And stood awhile in tmrwlmN, And, as in uffish thought he stood,   The Jabberwock, with eyes of flame, Came whiffling through the tulgey wood,   And burbled as it came! One, two! One, two! And through and throughaaaaaaavorpal blade went snicker-snack! He left it dead, and with its head   He went galumphing back. "And hast thou slain the Jabberwock?   Come to my arms, my beamish boy! O frabjous day! Callooh! Callay!"   He chortled in his joy. \'Twas brillig, and the slithy toves   Did gyre and gimble in the wabe: All mimsy were the borogoves,   And the mome raths outgrabe.'